core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java

```diff
@@ -0,0 +1,1793 @@
+      }
+      return returnEntry.getValue();
+   }
+
+   @Override
+   public V computeIfPresent(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+      Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED);
+      CacheEntryChange<K, V> ref = new CacheEntryChange<>();
+      return computeIfPresentInternal(key, remappingFunction, ref);
+   }
+
+   protected V computeIfPresentInternal(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction, CacheEntryChange<K, V> ref) {
+      boolean hasListeners = this.hasListeners;
+      getDataContainer().compute(key, (k, oldEntry, factory) -> {
+         V oldValue = getValue(oldEntry);
+         if (oldValue != null) {
+            V newValue = remappingFunction.apply(k, oldValue);
+            if (newValue == null) {
+               if (hasListeners) {
+                  cacheNotifier.notifyCacheEntryRemoved(k, oldValue, oldEntry.getMetadata(), true, ImmutableContext.INSTANCE, null);
+               }
+               ref.set(k, null, oldValue, oldEntry.getMetadata());
+               return null;
+            } else {
+               if (hasListeners) {
+                  cacheNotifier.notifyCacheEntryModified(k, newValue, defaultMetadata, oldValue, oldEntry.getMetadata(), true, ImmutableContext.INSTANCE, null);
+               }
+               ref.set(k, newValue, oldValue, oldEntry.getMetadata());
+               return factory.update(oldEntry, newValue, defaultMetadata);
+            }
+         } else {
+            return null;
+         }
+      });
+      V newValue = ref.getNewValue();
+      if (hasListeners) {
+         if (newValue != null) {
+            cacheNotifier.notifyCacheEntryModified(ref.getKey(), newValue, defaultMetadata, ref.getOldValue(), ref.getOldMetadata(), false, ImmutableContext.INSTANCE, null);
+         } else {
+            cacheNotifier.notifyCacheEntryRemoved(ref.getKey(), ref.getOldValue(), ref.getOldMetadata(), false, ImmutableContext.INSTANCE, null);
+         }
+      }
+      return newValue;
+   }
+
+   @Override
+   public V compute(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
+      CacheEntryChange<K, V> ref = new CacheEntryChange<>();
+      return computeInternal(key, remappingFunction, ref);
+   }
+
+   protected V computeInternal(K key, BiFunction<? super K, ? super V, ? extends V> remappingFunction, CacheEntryChange<K, V> ref) {
+      boolean hasListeners = this.hasListeners;
+      getDataContainer().compute(key, (k, oldEntry, factory) -> {
+         V oldValue = getValue(oldEntry);
+         V newValue = remappingFunction.apply(k, oldValue);
+         return getUpdatedEntry(k, oldEntry, factory, oldValue, newValue, ref, hasListeners);
+      });
+      return notifyAndReturn(ref, hasListeners);
+   }
+
+   @Override
+   public V merge(K key, V value, BiFunction<? super V, ? super V, ? extends V> remappingFunction) {
```

Comment:
We support it by default method, but it is pretty stupid (in this case it does a replace or putIfAbsent I believe).  Unfortunately Java 7/8 added a bunch of new methods which are prime for doing in a distributed/atomic way, but we haven't really made it a priority.